new-website/src/components/location/location-table.jsx

```diff
@@ -2,51 +2,93 @@ import React, { useState } from "react"
 import firebase from "gatsby-plugin-firebase"
 import MaterialTable from "material-table"
 import { LOCATION_COLLECTION } from "."
-import { useIntl } from "gatsby-plugin-intl"
+import { navigate, useIntl } from "gatsby-plugin-intl"
 import { generateI18N } from "./table-i18n"
 import { Roller } from "react-spinners-css"
+import shortid from "shortid"
 
-function choosePartnerId(entryPartnerId, userPartnerIds = []) {
-    if (entryPartnerId && userPartnerIds.includes(entryPartnerId)) {
-        return entryPartnerId
+function choosePartnerId(entryPartnerId, userPartnerIds, onError) {
+    if (entryPartnerId) {
+        if (userPartnerIds.includes(entryPartnerId)) {
+            return entryPartnerId
+        }
+        // TODO: Issue this error to sentry?
+        onError({ code: "choosePartnerId/inaccessible" })
+        throw new Error("Provided partner ID inaccessible for user")
+    } else {
+        // TODO: Depending on the context we may want users to choose and not just
+        // take the first in the sequence.
+        if (userPartnerIds.length > 0) {
+            return userPartnerIds[0]
+        }
+        // TODO: Issue this error to sentry?
+        onError({ code: "choosePartnerId/unavailable" })
+        throw new Error("No partner ID available for user")
     }
-    return userPartnerIds[0]
 }
 
-function addNewRow(userPartnerIds, onError) {
-    return async function add(data) {
-        if (!userPartnerIds || userPartnerIds.length <= 0) {
-            onError({ code: "no-permission" })
-            throw new Error("Missing Partner Id")
-        }
-        const { title, partnerId } = data
-        if (!title || title.length < 0) {
-            onError({ code: "requiredFieldMissing/title" })
-            throw new Error("missing title")
+function addNewRow(userPartnerIds, allPartnerConfigs, onError) {
+    const newPartnerLocationDoc = async () => {
+        // TODO: The partner name is a prefix in the location ID.
+        // Not sure this is ideal, but it is what we have currently.
+        const partnerId = choosePartnerId(null, userPartnerIds, onError)
+        const partnerConfig = allPartnerConfigs.find(p => p.id == partnerId)
+
+        // ShortID may collide with existing entry, retry a few times
+        for (let i=0; i<10; i++) {
+            const docId = `${partnerConfig.name}-${shortid.generate()}`
+            const docRef = firebase.firestore().collection(LOCATION_COLLECTION).doc(docId)
+            const doc = await docRef.get().catch(onError)
+            if (!doc.exists) {
+                docRef.set({ partnerId: partnerId })
+                return docId;
+            }
         }
+        onError({ code: "newPartnerLocationDoc/failed" })
+        throw new Error("Create new doc failed")
+    }
+
+    return async (event) => {
+        // For now, let's use a single page for add and edit
+        const docId = await newPartnerLocationDoc()
+        navigate(`/partner/location/${docId}`)
```

Comment:
The page for editing is still TODO. In case the user cancels the add operation we have to remember deleting the preliminary entry. However, I did it this way on purpose, because creating the new database entry here (and not when visiting the edit page), appears to give better control over the creation of locations. The edit page will only allow to change what exists already and not create arbitrary entries (i.e. that we would otherwise need to double-check there). What do you think?

Reply:
I would just send the user to the "add location" page and then do a error handling for collision on that page since operations on the firestore cost money 💸 so we should limit them as much as possible

Reply:
Well, yes, they will maybe cost money in the future. Thinking again, the other problem here is, that we may stick with ghost entries when we don't get an explicit `Cancel` action (user leaves the page, connection breaks, etc.). And while we could clean them up from time to time, we always had to consider them when querying the database. So, yes, I will change that.